File: app/src/main/java/com/niusounds/matomeviewer/util/VolleyUtils.java
package com.niusounds.matomeviewer.util;

import android.content.Context;
import android.graphics.Bitmap;
import android.support.v4.util.LruCache;

import com.android.volley.AuthFailureError;
import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.Response;
import com.android.volley.toolbox.ImageLoader;
import com.android.volley.toolbox.ImageRequest;
import com.android.volley.toolbox.StringRequest;
import com.android.volley.toolbox.Volley;

import org.androidannotations.annotations.AfterInject;
import org.androidannotations.annotations.EBean;
import org.androidannotations.annotations.RootContext;

import java.util.HashMap;
import java.util.Map;

@EBean(scope = EBean.Scope.Singleton)
public class VolleyUtils {
    private RequestQueue queue;
    private ImageLoader imageLoader;
    private Map<String, String> header = new HashMap<>();

    @RootContext
    Context ctx;

    @AfterInject
    void init() {
        queue = Volley.newRequestQueue(ctx);
        imageLoader = new ImageLoader(queue, new ImageLoader.ImageCache() {
            private final LruCache<String, Bitmap> cache = new LruCache<String, Bitmap>(20);

            @Override
            public Bitmap getBitmap(String url) {
                return cache.get(url);
            }

            @Override
            public void putBitmap(String url, Bitmap bitmap) {
                cache.put(url, bitmap);
            }
        });

        header.put("User-Agent", "Mozilla/5.0 (Linux; Android 4.2.1; en-us; Nexus 5 Build/JOP40D) AppleWebKit/535.19 (KHTML, like Gecko) Chrome/18.0.1025.166 Mobile Safari/535.19");
    }

    public <T> Request<T> request(String url, Class<T> clazz, Response.Listener<T> listener, Response.ErrorListener errorListener) {
        GsonRequest<T> request = new GsonRequest<>(url, clazz, null, listener, errorListener);
        queue.add(request);
        return request;
    }

    public Request<Bitmap> image(String url, Response.Listener<Bitmap> listener, int maxWidth, int maxHeight, Bitmap.Config decodeConfig, Response.ErrorListener errorListener) {
        ImageRequest request = new ImageRequest(url, listener, maxWidth, maxHeight, decodeConfig, errorListener);
        queue.add(request);
        return request;
    }

    public Request<String> requestString(String url, Response.Listener<String> listener, Response.ErrorListener errorListener) {
        StringRequest request = new StringRequest(url, listener, errorListener) {
            @Override
            public Map<String, String> getHeaders() throws AuthFailureError {
                return header;
            }
        };
        queue.add(request);
        return request;
    }

    public ImageLoader getDefaultImageLoader() {
        return imageLoader;
    }
}

File: app/src/main/java/com/niusounds/matomeviewer/util/VolleyUtils.kt
package com.niusounds.matomeviewer.util

import android.content.Context
import android.graphics.Bitmap
import android.support.v4.util.LruCache
import com.android.volley.Request
import com.android.volley.RequestQueue
import com.android.volley.Response
import com.android.volley.toolbox.ImageLoader
import com.android.volley.toolbox.StringRequest
import com.android.volley.toolbox.Volley
import org.androidannotations.annotations.AfterInject
import org.androidannotations.annotations.EBean
import org.androidannotations.annotations.RootContext

@EBean(scope = EBean.Scope.Singleton)
open class VolleyUtils {
    private val header = mapOf(Pair("User-Agent", "Mozilla/5.0 (Linux; Android 4.2.1; en-us; Nexus 5 Build/JOP40D) AppleWebKit/535.19 (KHTML, like Gecko) Chrome/18.0.1025.166 Mobile Safari/535.19"))

    private lateinit var queue: RequestQueue
    lateinit var defaultImageLoader: ImageLoader

    @RootContext
    lateinit var ctx: Context

    @AfterInject
    fun init() {
        queue = Volley.newRequestQueue(ctx)
        defaultImageLoader = ImageLoader(queue, object : ImageLoader.ImageCache {
            private val cache = LruCache<String, Bitmap>(20)

            override fun getBitmap(url: String): Bitmap? = cache.get(url)

            override fun putBitmap(url: String, bitmap: Bitmap) {
                cache.put(url, bitmap)
            }
        })
    }

    fun requestString(url: String, listener: Response.Listener<String>, errorListener: Response.ErrorListener): Request<String> {
        val request = object : StringRequest(url, listener, errorListener) {
            override fun getHeaders(): Map<String, String> = header
        }
        queue.add(request)
        return request
    }
}
